File: 216-combination-sum-iii/combination-sum-iii.py
class Solution(object):
    def combinationSum3(self, k, n):
        ans = []

        def solve(k, n, idx, temp):
            if k<0 or n<0:
                return
            if n==0 and k==0:
                ans.append(temp[:])
                return
            
            for i in range(idx,10):
                temp.append(i)
                solve(k-1, n-i, i+1, temp)
                temp.pop()
            
        solve(k,n, 1, [])
        return ans
            
                
        
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        

File: 216-combination-sum-iii/combination-sum-iii.java
class Solution {
    public void solve(List<List<Integer>> result, List<Integer> temp, int idx, int n, int k){
        if(temp.size()==k){
            if(n==0){
                result.add(new ArrayList<>(temp));
            }
            return;
        }
        for(int i = idx;i<10;i++){
            if(n<i) break;
            temp.add(i);
            solve(result, temp, i+1, n-i, k);
            temp.remove(temp.size()-1);
        }
        return;
    }
    public List<List<Integer>> combinationSum3(int k, int n) {
        int res = 0;
        List<List<Integer>> result = new ArrayList<>();
        for(int i=1;i<=k;i++){
            res+=i;
        }
        if(res>n) return result;
        List<Integer> temp = new ArrayList<>();
        solve(result, temp, 1, n, k);
        return result;
    }
}